File: lib/ios/MapView/MapViewManager.m
#import <React/RCTUIManager.h>

@interface RCT_EXTERN_MODULE(AMapViewManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(mapType, MAMapType)
RCT_EXPORT_VIEW_PROPERTY(initialCameraPosition, NSDictionary)
RCT_EXPORT_VIEW_PROPERTY(distanceFilter, double)
RCT_EXPORT_VIEW_PROPERTY(headingFilter, double)

RCT_REMAP_VIEW_PROPERTY(myLocationEnabled, showsUserLocation, BOOL)
RCT_REMAP_VIEW_PROPERTY(buildingsEnabled, showsBuildings, BOOL)
RCT_REMAP_VIEW_PROPERTY(trafficEnabled, showTraffic, BOOL)
RCT_REMAP_VIEW_PROPERTY(indoorViewEnabled, showsIndoorMap, BOOL)
RCT_REMAP_VIEW_PROPERTY(compassEnabled, showsCompass, BOOL)
RCT_REMAP_VIEW_PROPERTY(scaleControlsEnabled, showsScale, BOOL)
RCT_REMAP_VIEW_PROPERTY(scrollGesturesEnabled, scrollEnabled, BOOL)
RCT_REMAP_VIEW_PROPERTY(zoomGesturesEnabled, zoomEnabled, BOOL)
RCT_REMAP_VIEW_PROPERTY(rotateGesturesEnabled, rotateEnabled, BOOL)
RCT_REMAP_VIEW_PROPERTY(tiltGesturesEnabled, rotateCameraEnabled, BOOL)
RCT_REMAP_VIEW_PROPERTY(minZoom, minZoomLevel, double)
RCT_REMAP_VIEW_PROPERTY(maxZoom, maxZoomLevel, double)

RCT_EXPORT_VIEW_PROPERTY(onAPress, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onAPressPoi, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onALongPress, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onACameraIdle, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onACameraMove, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onALoad, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onALocation, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onACallback, RCTBubblingEventBlock)

RCT_EXTERN_METHOD(moveCamera:(nonnull NSNumber *)reactTag position:(NSDictionary *)_ duration:(int)_)
RCT_EXTERN_METHOD(call:(nonnull NSNumber *)reactTag callerId:(double)_ name:(NSString *)_ args:(NSDictionary *)_)

@end

File: lib/ios/MapView/MapViewManager.swift
@objc(AMapViewManager)
class AMapViewManager: RCTViewManager {
  override class func requiresMainQueueSetup() -> Bool { false }

  override func view() -> UIView {
    let view = MapView()
    view.delegate = view
    return view
  }

  @objc func moveCamera(_ reactTag: NSNumber, position: NSDictionary, duration: Int) {
    getView(reactTag: reactTag) { view in
      view.moveCamera(position: position, duration: duration)
    }
  }

  @objc func call(_ reactTag: NSNumber, callerId: Double, name: String, args: NSDictionary) {
    getView(reactTag: reactTag) { view in
      view.call(id: callerId, name: name, args: args)
    }
  }

  func getView(reactTag: NSNumber, callback: @escaping (MapView) -> Void) {
    bridge.uiManager.addUIBlock { _, viewRegistry in
      callback(viewRegistry![reactTag] as! MapView)
    }
  }
}

class MapView: MAMapView, MAMapViewDelegate {
  var initialized = false
  var overlayMap: [MABaseOverlay: Overlay] = [:]
  var markerMap: [MAPointAnnotation: Marker] = [:]

  @objc var onALoad: RCTBubblingEventBlock = { _ in }
  @objc var onACameraMove: RCTBubblingEventBlock = { _ in }
  @objc var onACameraIdle: RCTBubblingEventBlock = { _ in }
  @objc var onAPress: RCTBubblingEventBlock = { _ in }
  @objc var onAPressPoi: RCTBubblingEventBlock = { _ in }
  @objc var onALongPress: RCTBubblingEventBlock = { _ in }
  @objc var onALocation: RCTBubblingEventBlock = { _ in }
  @objc var onACallback: RCTBubblingEventBlock = { _ in }

  @objc func setInitialCameraPosition(_ json: NSDictionary) {
    if !initialized {
      initialized = true
      moveCamera(position: json)
    }
  }

  func moveCamera(position: NSDictionary, duration: Int = 0) {
    let status = MAMapStatus()
    status.zoomLevel = (position["zoom"] as? Double)?.cgFloat ?? zoomLevel
    status.cameraDegree = (position["tilt"] as? Double)?.cgFloat ?? cameraDegree
    status.rotationDegree = (position["bearing"] as? Double)?.cgFloat ?? rotationDegree
    status.centerCoordinate = (position["target"] as? NSDictionary)?.coordinate ?? centerCoordinate
    setMapStatus(status, animated: true, duration: Double(duration) / 1000)
  }

  func call(id: Double, name: String, args: NSDictionary) {
    switch name {
    case "getLatLng":
      callback(id: id, data: convert(args.point, toCoordinateFrom: self).json)
    default:
      break
    }
  }

  func callback(id: Double, data: [String: Any]) {
    onACallback(["id": id, "data": data])
  }

  override func didAddSubview(_ subview: UIView) {
    if let overlay = (subview as? Overlay)?.getOverlay() {
      overlayMap[overlay] = subview as? Overlay
      add(overlay)
    }
    if let annotation = (subview as? Marker)?.annotation {
      markerMap[annotation] = subview as? Marker
      addAnnotation(annotation)
    }
  }

  override func removeReactSubview(_ subview: UIView!) {
    super.removeReactSubview(subview)
    if let overlay = (subview as? Overlay)?.getOverlay() {
      overlayMap.removeValue(forKey: overlay)
      remove(overlay)
    }
    if let annotation = (subview as? Marker)?.annotation {
      markerMap.removeValue(forKey: annotation)
      removeAnnotation(annotation)
    }
  }

  func mapView(_: MAMapView, rendererFor overlay: MAOverlay) -> MAOverlayRenderer? {
    if let key = overlay as? MABaseOverlay {
      return overlayMap[key]?.getRenderer()
    }
    return nil
  }

  func mapView(_: MAMapView!, viewFor annotation: MAAnnotation) -> MAAnnotationView? {
    if let key = annotation as? MAPointAnnotation {
      return markerMap[key]?.getView()
    }
    return nil
  }

  func mapView(_: MAMapView!, annotationView view: MAAnnotationView!, didChange newState: MAAnnotationViewDragState, fromOldState _: MAAnnotationViewDragState) {
    if let key = view.annotation as? MAPointAnnotation {
      let market = markerMap[key]!
      if newState == MAAnnotationViewDragState.starting {
        market.onADragStart(nil)
      }
      if newState == MAAnnotationViewDragState.dragging {
        market.onADrag(nil)
      }
      if newState == MAAnnotationViewDragState.ending {
        market.onADragEnd(view.annotation.coordinate.json)
      }
    }
  }

  func mapView(_: MAMapView!, didAnnotationViewTapped view: MAAnnotationView!) {
    if let key = view.annotation as? MAPointAnnotation {
      markerMap[key]?.onAPress(nil)
    }
  }

  func mapInitComplete(_: MAMapView!) {
    onALoad(nil)
  }

  func mapView(_: MAMapView!, didSingleTappedAt coordinate: CLLocationCoordinate2D) {
    onAPress(coordinate.json)
  }

  func mapView(_: MAMapView!, didTouchPois pois: [Any]!) {
    let poi = pois[0] as! MATouchPoi
    onAPressPoi(["name": poi.name!, "id": poi.uid!, "position": poi.coordinate.json])
  }

  func mapView(_: MAMapView!, didLongPressedAt coordinate: CLLocationCoordinate2D) {
    onALongPress(coordinate.json)
  }

  func mapViewRegionChanged(_: MAMapView!) {
    onACameraMove(cameraEvent)
  }

  func mapView(_: MAMapView!, regionDidChangeAnimated _: Bool) {
    onACameraIdle(cameraEvent)
  }

  func mapView(_: MAMapView!, didUpdate userLocation: MAUserLocation!, updatingLocation _: Bool) {
    onALocation(userLocation.json)
  }
}
